### ISSUE TITLE
Where to ask questions regarding coravel?

### ISSUE THREAD

**Issue by pkellner** [jamesmh/coravel#226]
I don't see a. discussion forum  here or a reference to where there might be one.  If I missed it, my apologies.

The question I want to ask is whether using coravel supports a fixed, prioritized number of concurrent schedulej jobs.  That is, I have tasks that users perform from my web site an my goal is to queue those tasks, then wait on the result.  If my site gets hammered, I want to slow those tasks down from completing, but not kill the perf of my server by overloading it.

Am I making myself clear what I'm looking for?

Thanks for contributing this project, looks awesome.

**Comment by jamesmh** [comment]
Hey Peter, you're right - there's no official forum. I suppose the closest thing right now are the GitHub issues 😅.

Coravel has two different features around what I think you need: task scheduling and queuing. 

[Task scheduling](https://docs.coravel.net/Scheduler/) is a more static kind of thing - similar to having a bunch of cron jobs but defined in code.

There's also a [queuing feature](https://docs.coravel.net/Queuing/) - which I think is what you need. You can set how often the queue is processed/consumed (30 seconds by default). Right now the ability to prioritize specific tasks or a set a tasks (e.g. a named queue...) doesn't exist but is something that's been asked for.

Tasks that are queued are executed one after another - so having something like the ability to have multiple named queues that run concurrently would be, I think, what you need? But you'd also want some of those queues to be allocated more priority too right?

**Comment by pkellner** [comment]
The queuing is definitely what I'm looking for but with a multithread queue processor.  Years ago I started with the code in this project linked below to implement mostly what I'm looking for. 

That is, to have at something like 20 concurrent tasks that are pulled from a queue continually with a prioritization sign that  quick jobs can sneak on while long running lower priority one continue to be processed.  

https://www.codeproject.com/Articles/8886/Work-Queue-based-multi-threading

**Comment by jamesmh** [comment]
Make sense now. There's a piece of the task scheduling feature [that's very similar](https://docs.coravel.net/Scheduler/#schedule-workers) but without the priority semantics. Again, this is something that's been asked for quite a bit. I'll post on this issue when any decisions are made around this / work is done 👍

**Comment by pkellner** [comment]
I think the piece that I'm not seeing is the limit on how many workers will run at the same time.  In my case, I have 10,000 simulations I want to run.  I want to launch them all, but don't want more then say 25 running at the same time. I think now, if I launch them all with what you describe, they will all start running if I don't have the "preventOverlapping" set.

That's the scenario that would obviously collapse. My other scenario is that I may have hundreds of web users out there that all want to do work that is too long to do in a request.  I want that work queued, then the request can either poll or make a signalR type connection to keep track of when it is completed.  In this case, I can limit how much memory gets tied up and not overwhelm the server.

**Comment by pkellner** [comment]
Also, for me, the prioritization is not that important, as I think the multiple queues solves that.  In my case, I just have a few different priorities and having a few queues is very managable.

**Comment by pkellner** [comment]
Sorry for the multiple "stream-of-thought" type issues here ...

At this URL https://docs.coravel.net/Queuing/#metrics you mention "waiting jobs".  How does a job end up waiting? I can't find any docs that mention queuing something that does not start running immediately.

**Comment by jamesmh** [comment]
Np 😀 Queued jobs will "wait" because item's in the queue are processed every X seconds (polling). [This is configurable](https://docs.coravel.net/Queuing/#adjusting-consummation-delay).

If someone decides to only process the queue every 60 seconds, then it will take a little longer for their jobs to complete as they are "waiting" before being released/processed.

Hopefully that makes sense?

**Comment by pkellner** [comment]
Yes, that makes sense. I had not thought about the periodic processing.  I'm going to try Coravel in my new project.  Seems like a lot of very useful stuff that's come from real world needs.  I think if I really want the worker item multi-threaded processing, I can likely do that separately in my app as it's not that hard to manage.  This is the first time in a while I've needed that kind of processing of worker items. I was hoping to find it built out there better then I would do it myself.

**Comment by jamesmh** [comment]
Sounds good 🙂 Don't hesitate to let me know how it goes or any other questions 👍
